docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md

```diff
@@ -409,3 +411,34 @@ terraform {
   source = "${include.env.locals.source_base_url}?ref=v0.2.0"
 }
 ```
+
+### Considerations for CI/CD Pipelines
+
+For infrastructure CI/CD pipelines, it is common to only want to run the workflow on the modules that were updated. For
+example, if you only changed the `terragrunt.hcl` configuration for the RDS database in the dev account, then you only
+want to run `plan` and `apply` on that module, not other components or other accounts.
+
+If you did not take advantage of `include` or `read_terragrunt_config`, then implementing this pipeline is
+straightforward: you can use `git diff` to collect all the files that changed, and for those `terragrunt.hcl` files that
+were updated, you can run `terragrunt plan` or `terragrunt apply` by passing in the updated file with
+`--terragrunt-config`.
+
+However, if you use `include` or `read_terragrunt_config`, then a single file change may need to be reflected on
+multiple files that were not touched at all in the commit. In our previous example, when a configuration is updated in
+the `_env/app.hcl` file, we need to apply the change to all the modules that `include` that common environment
+configuration.
+
+Terragrunt currently does not have any features for supporting this use case when `read_terragrunt_config` is
+used. However, for `include` blocks, you can use the
+[--terragrunt-modules-that-include]({{site.baseurl}}/docs/reference/cli-options/#terragrunt-modules-that-include) CLI
+option for the `run-all` command.
+
+In the previous example, your CI/CD pipeline can run `terragrunt run-all plan --terragrunt-modules-that-include
+_env/app.hcl`. This will:
+
+- Recursively find all Terragrunt modules in the current directory tree.
+- Filter out any modules that don't include `_env/app.hcl` so that they won't be touched.
+- Run `plan` on any modules remaining (which will be the set of modules in the current tree that include
```

Comment:
How does `apply` work in this case? In particular:

1. If multiple modules in a single env include something that was modified: e.g., you have some `_env/db.hcl` and you use that 5 times in the `dev` environment to deploy 5 DBs. Will / should our CI / CD pipeline try to `apply` them all in parallel? I'm guessing the answer is yes.
2. If multiple modules across different envs include this thing: e.g., the `_env/db.hcl` is used in `dev`, `stage`, and `prod`. Will / should our CI / CD pipeline try to `apply` them all in parallel? I'm guessing the answer here should be no. But then, is our pipeline able to run `plan` in one env, prompt for approval, if approved, run `apply`, then automatically repeat the process with each subsequent env?

Reply:
I mentioned this in the PR description, but realized I didn't do it for the docs so expanded on it: 9c4dc14b3fae49d361e230796c3890420e353dec

Note that I think this is still flawed, in that the commit is a one way street. That is, if you abort after deploying to `stage`, you are now in an awkward state where you have to roll this change out when you want to touch `prod` since the updates are already merged for `qa` and `stage`.

I think the real solution here is going to be figuring out how to implement remote `include`, although that has its own can of worms as we discussed.

Reply:
Ah, yea, you're right. This would require a rollback if the deploy failed part way through. I think the best we can do for now is document the need for rollback, both in the Terragrunt docs, and our own pipeline docs. In the future, as you said, remote `include` is prob the proper solution.

Reply:
Good point about documenting. What do you think about adcafb8f3b8656794301dec00c90355c94f3bdf6?

Reply:
LGTM